clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -420,39 +420,39 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
       // passed to the parallel_for function). For all other cases,
       // emit a warning and ignore.
       if (auto *A = FD->getAttr<SYCLIntelKernelArgsRestrictAttr>()) {
-        if (ParentFD == SYCLKernel) {
+        if (ParentFD == SYCLKernel && A->isImplicit()) {
```

Comment:
I would think you would want this logic reversed, no?  Why would we collect only the ones that were added implicitly?  Won't this prevent directly called function attributes to be skipped?

Reply:
Yes. This would prevent directly called function attributes to be skipped. Also isImplicit() shouldn't be used at all. It causes skipping all attributes that are explicitly added by the user, irrespective of whether they are direct or indirect calls.